### DISCUSSION TITLE
How to access browser location url query string on server?

### DISCUSSION THREAD

**Discussion by earlsioson** [Kitware/trame#83] [Q&A] [answered]
How would I access the http request querystring parameters on the server? Do we have access to an http request object for the current route?

**Comment by jourdain** [comment, marked as answer]
You can do something like [that](https://github.com/Kitware/trame/discussions/70#discussioncomment-2836193)

**Reply by earlsioson** [reply]
Thank you again, per your link, I think doing something like the following should work for me.

```python
from trame.app import get_server
from trame.ui.html import DivLayout
from trame.widgets import trame

server = get_server()
state, ctrl = server.state, server.controller

layout = DivLayout(server)


def get_search():
    print("#######", state.search)
    return state.search


with layout:
    trame.ClientTriggers(mounted="search = window.location.search")
    trame.ClientStateChange(value="search", change=ctrl.get_search)
    ctrl.get_search = get_search

state.update(
    {
        "search": None,
    }
)

if __name__ == "__main__":
    server.start()

```

**Reply by jourdain** [reply]
I don't think that will actually work but you can do something simpler like

```python
@ctrl.set("get_search")
def get_search(search):
   print(search)

trame.ClientStateChange(value="window.location.search", change=(ctrl.get_search, "[window.location.search]"))
```

**Reply by earlsioson** [reply]
@jourdain Thank you your example worked for us

**Reply by chriswallis** [reply]
Hi there.  Would it be possible to elaborate on your example, please?  I've tried adapting the [earlsioson](https://github.com/earlsioson) code above with your example, to give:

```
from trame.app import get_server
from trame.ui.html import DivLayout
from trame.widgets import trame

server = get_server()
state, ctrl = server.state, server.controller

@ctrl.set("get_search")
def get_search(search):
   print(search)

layout = DivLayout(server)

with layout:
    trame.ClientStateChange(value="window.location.search", change=(ctrl.get_search, "[window.location.search]"))

if __name__ == "__main__":
    server.start()
```

I then visit my URL, which looks like:  http://localhost:1234/index.html?foo=bar

Nothing is printed to output (I've also tried writing to file) and I don't see a ws message containing the data, either.

I think the problem may be that this does not fire on initial page load, but that is when we need it.

Many thanks

**Reply by jourdain** [reply]
So if you just want to get the execution when the app loads, you just need to do something like

`trame.ClientTriggers(mounted=(ctrl.get_search, "[window.location.search]"))`

**Comment by wholmen** [comment]
@jourdain  I can't import trame from widgets anymore. Is there an updated way of doing this? 

I'm trying to embedd a trame-application in a larger React-application, and getting hold of query-parameters/path-parameters is crucial for this setup to work. 

Edit: I hadn't installed trame-components, so I could find them now. 

I also have some trouble building an application that depend on the query parameters. E.g. for authentication and db-fetching. Could you help me figure out how to use the query parameters that is extracted in the get_search-method and render different pages depending on what's in the query-parameters? 

I would want to set up something like: 

```python
from typing import Any

from trame.app import get_server
from trame.ui.vuetify3 import SinglePageLayout
from trame.widgets import trame
from trame_server.controller import Controller
from trame_server.state import State

server = get_server()
server.client_type = "vue3"

state: State = server.state
ctrl: Controller = server.controller

layout = SinglePageLayout(server)


@ctrl.set("get_search")
def get_search(search):
    if search.startswith("?"):
        search = search[1:]
    query_list = search.split("&")
    queries = [(item.split("=")[0], item.split("=")[1]) for item in query_list]
    state.search = {
        "token": [query_tuple[1] for query_tuple in queries if query_tuple[0] == "token"][0],
        "foo_id": [query_tuple[1] for query_tuple in queries if query_tuple[0] == "foo_id"][0],
    }
    return state.search


def render_app(layout: SinglePageLayout):
    pass


with layout:
    trame.ClientTriggers(mounted=(ctrl.get_search, "[window.location.search]"))

    if hasattr(state, "search") and isinstance(state.search, dict):
        token = state.search.get("token", None)
        foo_id = state.search.get("foo_id", None)
        
        if token and foo_id:
            render_app(layout)
        elif token and not foo_id:
            layout.title.set_text("Couldn't find foo")
        elif not token and foo_id:
            layout.title.set_text("Couldn't authenticate")
        else:
            layout.title.set_text("Couldn't authenticate")
    else:
        layout.title.set_text("Loading")
```

**Comment by wholmen** [comment]
I solved it by putting the application inside the get_search, function. But I have the same problem as chriswallis. The line
```python 
trame.ClientStateChange(value="window.location.search", trigger_on_create=True, change=(ctrl.get_search, "[window.location.search]"))
```

Doesn't seem to do anything when updating the url parameters. 

I need the server to respond to changing requests, so it must be able to updating on a change in query parameters

**Reply by jourdain** [reply]
There might be a way to leverage the reactivity of the router to automatically detect change.

**Comment by scutlemon** [comment]
I used the following code, which can retrieve the requested parameters, but cannot retrieve the data immediately. The previous data will not be retrieved until the next entry. How to solve this?
![image](https://github.com/user-attachments/assets/8f45c03d-50af-45fe-a636-cc3dbdeab8da)


code :
from trame.app import get_server
from trame.ui.html import DivLayout
from trame.widgets import trame

server = get_server()
state, ctrl = server.state, server.controller

layout = DivLayout(server)

@ctrl.set("get_search")
def get_search():
    print("#######", state.search)
    return state.search

with layout:
    trame.ClientTriggers(mounted="search = window.location.search")
    trame.ClientStateChange(value="search", change=ctrl.get_search, immediate=True, trigger_on_create=True )

    #trame.ClientStateChange(value="window.location.search", change=(ctrl.get_search, "[window.location.search]"))
    #trame.ClientTriggers(mounted=(ctrl.get_search, "[window.location.search]"))

    ctrl.get_search = get_search

state.update(
    {
        "search": "0",
    }
)

if __name__ == "__main__":
    server.start()
